### ISSUE TITLE
Mypy can't tell if a variable can't be False outside a loop

### ISSUE THREAD

**Issue by faresbakhit** [python/mypy#11333]
Sorry for the unclear title. Please edit it.

**Bug Report**

In the following snippet the only conditions to break out of the loop is either
that `default` wouldn't be `None` or `value` has one or more characters. I assume
that Mypy is reporting that error because of the latter so it can't tell that
(or ...) would be unreachable if the loop finished because `value` is `True`.

```python
from typing import Optional


def ask(question: str, default: Optional[str] = None) -> str:
    while not (value := input(question)):
        if default is not None:
            break
    return value or default
```

```
error: Incompatible return value type (got "Optional[str]", expected "str")
```

**Your Environment**

- Mypy version used: 0.910
- Python version used: 3.9
- Operating system: Linux

**Comment by hauntsaninja** [comment]
mypy isn't capable of understanding relationships between variables like this and it's not something that it will be capable of any time soon, unfortunately.
